rfcs/0000-buildpack-resource.md

```diff
@@ -0,0 +1,218 @@
+
+##### Builder Resources
+
+(Cluster)Builder resources do not need to explicitly reference Buildpack resources.
+
+An example Builder without a configured ClusterStore is shown below.
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Builder
+metadata:
+  name: my-builder
+spec:
+  tag: gcr.io/sample/builder
+  stack:
+    name: bionic-stack
+    kind: ClusterStack
+  order:
+  - group:
+    - id: paketo-buildpacks/java      
+```
+The `paketo-buildpacks/java` buildpack referenced in the Builder order would be sourced from the Buildpack resource within the `my-builder` namespace or the ClusterBuildpack resource that provides the `paketo-buildpacks/java` with the highest version number as defined by semver. Buildpacks IDs defined in the builder order may be a "implementation buildpack" within a meta-buildpack. If multiple Buildpack resources provide the same selected Buildpack the Builder should reconcile with an error.
+
+Builders should also be able to reference a Buildpack resource directly with an object reference within their order definition. An example Builder is shown below with an object reference within its order.
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Builder
+metadata:
+  name: my-builder
+spec:
+  tag: gcr.io/sample/builder
+  stack:
+    name: bionic-stack
+    kind: ClusterStack
+  order:
+  - group:
+    - id: paketo-buildpacks/java
+    - name: my-custom-app-buildpack-0.3.1
+      kind: Buildpack
+```
+
+Builders may also reference a buildpack within a buildpack resource and not necessarily the entire buildpackage. This is possible by providing an ID alongside an object reference.
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Builder
+metadata:
+  name: my-builder
+spec:
+  tag: gcr.io/sample/builder
+  stack:
+    name: bionic-stack
+    kind: ClusterStack
+  order:
+  - group:
+    - name: my-custom-app-buildpack-0.3.1
+      kind: Buildpack
+      id: my-custom/buildpack-inside
+```
+
+Alternatively, If users do not desire the complexity of external (Cluster)Buildpack resources, buildpackage images can be referenced directly within the Builder order. This removes the need to extenalize buildpack references from the Builder resource:
+
+```yaml
+apiVersion: kpack.io/v1alpha3
+kind: Builder
+metadata:
+  name: my-builder
+spec:
+  tag: gcr.io/sample/builder
+  stack:
+    name: bionic-stack
+    kind: ClusterStack
+  order:
+  - group:
+    - image: gcr.io/paketo-buildpacks/java:5.9.1
+    - name: my-custom-app-buildpack-0.3.1
+      kind: Buildpack
+```
+
+
+**Prior Art:**
+* [kapp controller](https://carvel.dev/kapp-controller/) models available Package versions as independent Package resources.
+
+**Risks:**
+This will require introducing a new api version within kpack. Future kpack releases will need to maintain support for ClusterStore resources until the kpack apis which contain the ClusterStore are removed.
```

Comment:
This is really great question. I was assuming that ClusterStores can be used similar to ClusterBuildpacks as essentially a large deprecated ClusterBuildpack. Do you think this makes sense or the ClusterStore should be reference-able directly?

Reply:
> Moreover, can we translate the old resource to the new one instead of having them both live side by side for a few releases?
I don't believe we have an easy way to translate them. Existing cluster with kpack installed may have the resources saved into etcd. We would still need to support those resources.

Reply:
I think the best migration path is:
- v1alpha3:
  - ClusterStores can be used in builders and clusterbuilders via store key
  - If store key is unset Buildpacks and ClusterBuildpacks are used
- v1alpha4
  - ClusterStores are removed entirely

Reply:
@tylerphelan, regarding your suggestion about v1alpha3:
Are you ok that ClusterStores will be used even if store key is set?

Also, what will happen in v1alpha4 with the ClusterStores? It's a breaking change. What will be the process of upgrading to this version?

Reply:
> Are you ok that ClusterStores will be used even if store key is set?
I think this provides a migration path for the api change

> Also, what will happen in v1alpha4 with the ClusterStores? It's a breaking change. What will be the process of upgrading to this version?

I think this will be a true breaking change. We can keep the legacy behavior if we want but I supposed at some api version we should remove it.

Reply:
> Are we sure it'll be a big problem to make a translation from ClusterStores to ClusterBuildpacks?

I think attempting to map and migrate existing ClusterStores to ClusterBuildpacks could be useful but, It doesn't seem like there is an obvious strategy for how to do the mapping. Any large api change like this needs to allow sometime for users to migrate so, I don't believe we could remove support for ClusterStores. I think *if and when* Clusterstores are formally removed would be a good time to explore migration tooling. Is it okay if we leave that outside the scope of this RFC?